src/main/java/com/aphyr/riemann/client/IPromise.java

```diff
@@ -28,6 +29,10 @@ public Object unsafeDeref(final long time,
                             final TimeUnit unit,
                             final Object timeoutValue) throws IOException;
 
+
+  // Set a callback for this promise; will be invoked when the promise delivers
+  public void setCallback(Callback<T> c);
```

Comment:
The reason I designed the promise system without callbacks was to avoid problems with concurrent state and mutation. Having a single callback means you need to guarantee (and at runtime, since you can't enforce it statically) that each promise only has a single callback assignment. Otherwise you're in the awkward position of losing callbacks when someone else assigns them.